exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,561 @@
+use jni::{
+    objects::{GlobalRef, JObject, JValue},
+    signature::{JavaType, Primitive},
+    Executor, JNIEnv
+};
+use proto;
+use proxy::node::NodeContext;
+use runtime::Error;
+use std::fmt;
+use storage::View;
+use to_handle;
+use utils::{
+    jni_cache::{classes_refs, runtime_adapter},
+    panic_on_exception, unwrap_jni, get_and_clear_java_exception,
+    describe_java_exception, get_exception_message,
+};
+use JniErrorKind;
+use JniResult;
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+}
+
+impl JavaRuntimeProxy {
+    /// Runtime Identifier
+    pub const RUNTIME_ID: RuntimeIdentifier = RuntimeIdentifier::Java;
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            Err(Error::IncorrectArtifactId.into())
+        } else {
+            Ok(JavaArtifactId(artifact.name.to_string()))
+        }
+    }
+
+    /// If the current node is a validator, returns `Some(validator_id)`, for other nodes return `None`.
+    fn validator_id(snapshot: &dyn Snapshot, pub_key: &PublicKey) -> Option<ValidatorId> {
+        CoreSchema::new(snapshot)
+            .consensus_config()
+            .find_validator(|validator_keys| *pub_key == validator_keys.service_key)
+    }
+
+    /// If the current node is a validator, return its identifier, for other nodes return `default`.
+    fn validator_id_or(
+        snapshot: &dyn Snapshot,
+        pub_key: &PublicKey,
+        default: i32
+    ) -> i32 {
+        Self::validator_id(snapshot, pub_key)
+            .map_or(default, |id| i32::from(id.0))
+    }
+
+    /// Converts `JniResult<R>` into `Result<R, ExecutionError>`.
+    fn convert_jni<R>(res: JniResult<R>) -> Result<R, ExecutionError> {
+        res.map_err(|err| {
+            let kind: ErrorKind = match err.kind() {
+                JniErrorKind::JavaException => Error::JavaException.into(),
+                _ => Error::OtherJniError.into(),
+            };
+            (kind, err).into()
+        })
+    }
+
+    /// Handles and clears any Java exceptions.
+    ///
+    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
+    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
+    /// being thrown, then exception is passed to corresponding `ExceptionHandler` according their
+    /// type and `exception_handlers` mapping.
+    /// `ExceptionHandlers::DEFAULT` is called in case of there is no any handlers or handlers are
+    /// not matched to exception type.
+    fn handle_exception<H, R>(
```

Comment:
So now we have three methods `convert_jni`, `handle_exception` and `handle_any_exception` with pretty similar semantics and purpose. Can we reduce it?